Having a try at post body as form-url encoded

AnthonySteele opened the pull request:
We need to send some posts as form-url-encoded data
This should allow it. Will test more thoroughly after lunch
Code from http://ole.michelsen.dk/blog/serialize-object-into-a-query-string-with-reflection.html

Advantage of form-url-encoded data is that the body is used in the OAuth signing, so it can't be changed without the request being rejected.

AnthonySteele commented:
I don't know yet if the oauth is working, hold off on merging until this is confirmed

AnthonySteele commented:
Right, this is ready fro review.

We have verified that this works now - the body has to be decoded and put into the params to OAuth signing (in RequestBuilder).

It's not ideal that RequestBuilder has to use HttpUtility.ParseQueryString to turn the body (string) back into params, but we have verified that it work with a signed POST endpoint that only allows  form-urlencoded data (because they explicitly wanted the body data to be OAuth signed to resist tampering to that part of the message)

gregsochanik commented:
Couple of things:

`SerialiseWithRefection` should read `SerialiseWithReflection` (missing l in Reflection)

Also, do we have to have those in-line comments in the private method here? I can see the value of the summary re credit/further reading, but not the in-line comments.

Also - a couple more tests in `FormUrlEncodedPayloadSerializerTests` wouldn't go amiss, just to show:

* That null vals are ignored
* That nested DTOs (i.e. Artist in Release) serializes to its respective ToString(). 

Other than that looks good; if you're happy with comma separated string as a representation for IEnumerables then so am I!

I guess `IPayloadSerializer` ties us to at least having to give a nod to representing complex types in a limited key/value pair string, even if it's less than ideal.

AnthonySteele commented:
I have fixed the typo and removed the in-line comments.

Regarding the output of serialization, the proof of the pudding is in the eating - if nancy deserialises it back to the same data then we're doing OK. I'll put in more tests - is there a simple way to do a round-trip test? especially for the list of ints (IEnumerables), which is what we actually need for an endpoint.

Regarding  representing complex types there's no easy answer. If you really have complex data then you should consider a json or xml body instead of form-url, but then you can't do oauth signing including that body's data :(

gregsochanik commented:
Not sure how easy a round trip test would be, not sure it's worth it for this PR. Might be worth experimenting once it's merged.

Do we have any endpoints that actually deserialise form-urlencoded that we can get a assertable response from?

AnthonySteele commented:
Yep. I have verified that we can talk to the endpoint (meaning that oauth worked), I will test tomorrow that it gets the right data out of a serialised list of int

AnthonySteele commented:
I have also confirmed in the debugger than nancy correctly de-serialised this data back into a list of ints.
Anything else?

AnthonySteele commented:
Suggested release notes:

Handle turning an object into a `POST` body in `application/x-www-form-urlencoded` format. 
This format has the limitation that it is a collection of `name=value` pairs and so is not suitable for complex or deeply nested objects – for these cases we already support json and xml post bodies.  But it is preferred by some `POST` endpoints where security is important, as the names and values are fed into the Oauth signature, so changes to the form-urlencoded data will invalidate the request's authorisation.


e.g to produce a form-url encoded body like `“Id=42&Name=Fred”` you would do:

```
var data = new EndpointPayload { Id = 42, Name = “Fred” };

var request = _api.Create<SomeEndpoint>()
	.WithPayload(data, PayloadFormat.FormUrlEncoded);
```

gregsochanik commented:
Can't think of anything else, release notes look good. Let me know when you're happy and I'll merge.

AnthonySteele commented:
Only other thing - the tests in `RequestBuilderOAuthHeaderTests` on `RequestBuilder` are useful to step into RequestBuilder and see what it's doing, but don't test the details of e.g. then the post body is read and when it is ignored. Not sure what else we can do here without larger changes?

Other than that, we think that it works, you can merge.
